### ISSUE TITLE
Change FID to be online

### ISSUE THREAD

**Issue by nicolas-dufour** [Lightning-AI/metrics#1198]
## 🚀 Feature

The idea is to compute mean and covariance online after each update instead of storing the features and then computing the mean and covariance

### Motivation

The motivation is to avoid storing the features which will lead to a way more memory efficient metric.

### Pitch

This can be done the following way:
```python
def update(self, imgs: Tensor, real: bool) -> None:  # type: ignore
        """Update the state with extracted features.

        Args:
            imgs: tensor with images feed to the feature extractor
            real: bool indicating if ``imgs`` belong to the real or the fake distribution
        """
        features = self.inception(imgs).double()
        self.orig_dtype = features.dtype

        if real:
            self.real_features_sum += features.sum(dim=0)
            self.real_features_cov_sum += features.t().mm(features)
            self.real_features_num_samples += imgs.shape[0]
        else:
            self.fake_features_sum += features.sum(dim=0)
            self.fake_features_cov_sum += features.t().mm(features)
            self.fake_features_num_samples += imgs.shape[0]

    def compute(self) -> Tensor:
        """Calculate FID score based on accumulated extracted features from the two distributions."""
        mean_real = (self.real_features_sum / self.real_features_num_samples).unsqueeze(0)
        mean_fake = (self.fake_features_sum / self.fake_features_num_samples).unsqueeze(dim=0)

        cov_real = (self.real_features_cov_sum / self.real_features_num_samples - mean_real.t().mm(mean_real)) / (
            self.real_features_num_samples - 1
        )
        cov_fake = (self.fake_features_cov_sum - self.fake_features_num_samples * mean_fake.t().mm(mean_fake)) / (
            self.fake_features_num_samples - 1
        )
        return _compute_fid(mean_real, cov_real, mean_fake, cov_fake).to(self.orig_dtype)
```

**Comment by SkafteNicki** [comment]
Hi @nicolas-dufour,
Thanks for raising this issue. Feel free to send a PR with your proposed change as it seems like an awesome enhancement :)

**Comment by kimihailv** [comment]
@nicolas-dufour But this implementation depends on batch size. For instance, if we pass `imgs` with shape n x c x h x w, we will get statistics which will be differ from statistics if we iteratively (by batch) pass the same `imgs`

**Comment by nicolas-dufour** [comment]
@kimihailv I don't get what you are saying. This implementation is not batch dependent as we divide by the batch size.
The reasoning behind this is that both the covariance matrix and the mean can be computed by a running average.

if $| X| $ the cardinal of X,
For the mean:

$\mu_X = \sum_{x\in X} x / |X| =  \sum_{i \in [|1;n|]} \sum_{x\in X[i:i+batch_size]} x / |X| $

For the covariance:

$\Sigma_X = \sum_{x\in X} (x- \mu_X ) (x- \mu_X )^T/ |X| =  \sum_{x\in X} xx^T/ |X|  -  \sum_{x\in X}\mu_X \mu_X^T/ |X| $
$=  \sum_{i \in [|1;n|]} \sum_{x\in X[i:i+batch_size]} xx^T/ |X| -  \sum_{i \in [|1;n|]} \sum_{x\in X[i:i+batch_size]} \mu_X \mu_X^T / |X|$

As you can see there is no batch dependence in this formula and if i'm not wrong this is what's implemented in this PR.

**Comment by nicolas-dufour** [comment]
@kimihailv after investigating, i've understood what you meant. The FID is different if we use different batch-sizes. After experimenting, i found this to be linked to the inception network and not the FID computation. Indeed, if we change 
```
features = self.inception(imgs)
```

by 

```
features = []
        for img in imgs:
            feature = self.inception(img.unsqueeze(0))
            features.append(feature)
        features = torch.cat(features, dim=0)
```
now the FID is batch independent again (same results regardless of the batch). Sadly, this seems to be a very inefficient way to compute the features.
It's super weird that the underlying tensorflow network change it's output depending of the batch-size.

 I think that we should try to compute the features with float.64, but this would require some changes with torch fidelity which seems to be stale since 2021. 
 I will open a new issue with this problem since it is not related to the online implementation.
